### ISSUE TITLE
One EventLoop per ExecutionContext

### ISSUE THREAD

**Issue by ysbaddaden** [ysbaddaden/execution_context#7]
For compatibility with the current Crystal stdlib, execution contexts create one EventLoop per thread. This is mostly for IO::Evented that keeps an Event for each thread.

We could try to have one event loop per execution context instead.

**Pro**: it would limit the requirement to keep a list of schedulers blocked on the event loop, but allow to have a single thread blocked (one atomic vs lock + linked list). With M:N schedulers and threads we could even try to follow on Go and detach the scheduler from the thread and let it block on the event loop, instead of having to detect & interrupt the event loop.

**Con**: we could potentially have a contention point when creating events (especially with lots of threads).

**Comment by ysbaddaden** [comment]
**Pro**: with a single Event Loop (EL) per EC, whenever a scheduler has nothing to do it could try to run the EL and trigger all the activable events for the whole execution context.

With one EL per scheduler, the scheduler would only handle the activable events that happened to be in its EL, while other activable events (in the other schedulers' EL) will only be processed when that specific scheduler has nothing to do, which means we risk some activable events to be delayed, possibly for a long time, and possibly multiple times.

For example a CPU-bound fiber may run for some time (milliseconds, maybe seconds), and that will block the events that happen to be in that thread's EL for that duration, increasing delays in response times.

That blocked event may be the signal handler loop, that processes SIGCHLD which is how we detect that a spawned process terminated, which means that a CPU bound fiber could block other fibers from doing any progress until _one_ fiber yields.

**Comment by ysbaddaden** [comment]
Trying to run only one EventLoop per ExecutionContext in #29 sadly ruins the MT performance, where MT:4 goes from 170k down to 67k req/s (same as MT:1). I assume the mutex in Crystal::Evented forces too many thread synchronization.

- we lock on each enqueue and dequeue (IO, timer);
- we lock for the whole duration of processing ready events/timers;

This means that we can't enqueue or dequeue while another thread is processing events, in addition to other threads trying to wait.

Of course this is but a silly benchmark, yet it still highlights the contention issue.

**Comment by ysbaddaden** [comment]
Enabling traces, I see that it took ~80us to process 17 events for a thread after waiting (holding the lock) during which _nothing_ else happens, yet immediately after the 3 other threads add a fd to epoll.

**Comment by ysbaddaden** [comment]
Looking more broadly to the traces outlined that one thread does 99% of the work, while the others did things for the first 1% then stopped doing anything, which explains why we go down to MT:1 performance: only one thread is running, the others parked themselves.

Now, to understand _why_.

**Comment by ysbaddaden** [comment]
I fixed the issue. Performance is down to 135k req/s. That's twice better, but still below 80% of the original performance of one EL per thread.

**Comment by ysbaddaden** [comment]
This has been implemented.
